### PR TITLE
[WIP] Feature/branch search box

### PR DESCRIPTION
**Pull request by ZimGil**
I was able to make the search box operate as expected in #1196 

Some bugs that I would like to get some help with:
 * ~~After choosing a branch for checkout, the search string stays but the branches are added again into the list.~~
 * ~~I still iterate twice over some branches (once for all the available and 2nd time over the ones that was searched)~~
 * I could not figure out how to change the color of the `placeholder` from white, now it looks like an empty search box until user input.
 * I could not find a way to avoid iterating over the available branches (`branchesAndLocalTags`) in order to fill the `filteredBranchesAndLocalTags` when the `searchString` is empty.
    > Tried using: `this.filteredBranchesAndLocalTags(this.branchesAndLocalTags());` but then the search got stuck.
    > All other methods (that I've found) to fill an observable array do it by iterating.

I've never worked with Knockout.js so I would like to see where I can make this better.
@jung-kim @ylecuyer

### CODE REVIEW

File: components/branches/branches.js [hunk 1]
```diff
@@ -17,6 +17,11 @@ class BranchesViewModel {
     this.repoPath = repoPath;
     this.server = server;
     this.branchesAndLocalTags = ko.observableArray();
+    this.filteredBranchesAndLocalTags = ko.observableArray();
+    this.branchesAndLocalTags.subscribe((branches) => {
```
Comment (ZimGil) [inline components/branches/branches.js]: > * After choosing a branch for checkout, the search string stays but the branches are added again into the list.

Oh now I see that its actually happening on every `git pull`

**Comment by jung-kim** [comment]
- Ermmm, I'm not sure on this one either myself and front-end is a never ending mystery for me.  but I thought placeholder will inherit text color no?
- I see what you are trying to do here, and yes you would have to iterate over all branches everytime search string changes.  Infact, if you see the examples I have shared with you before this is all done via [jquery autocomplete](https://jqueryui.com/autocomplete/) library.  Is there a particular reason why we are not using this lib

**Comment by ylecuyer** [comment]
I haven't read the code yet but tested the feature instead:

1. As you said in the description it is all white, white background, white input, white placeholder, maybe you can add a subtle grey as a background and replace the place holder with a search icon on the right maybe?
2. the debounce doesn't feel ok. I understand it may be needed if there is a lot of branches but for me it looked awkward. For instance, with the ungit repo, if you search for `0/` you can have results though there is no match
3. I tried without debounce and to fully clear the search, you have to press the back key twice. eg: search 0.1 then back / back one more time back and you wont have the full list, back once again and you have the full list.
4. Instead of populating the list of branches and remotes, maybe you can add a flag isHidden to each branch/remote ? it will speedup the process and avoid duplication

File: components/branches/branches.html [hunk 2]
```diff
@@ -22,7 +22,11 @@
       </label>
     </div>
     <div class="dropdown-divider"></div>
-    <!-- ko foreach: branchesAndLocalTags -->
+    <div class="form-inline" onclick="event.stopPropagation()">
+      <input class="form-control branch-search-box" type="text" placeholder="Search Branchs" data-bind="value: branchSearchString, valueUpdate: 'afterkeydown'"/>
```
Comment (ylecuyer) [inline components/branches/branches.html]:
```suggestion
      <input class="form-control branch-search-box" type="text" placeholder="Search Branchs" data-bind="value: branchSearchString, valueUpdate: 'afterkeyup'"/>
```

should fix my comment 3.

**Comment by ZimGil** [comment]
* I don't really like the auto complete solution, you'll end-up having the full list and the auto complete coveting it in some way. I like the idea of having an `isHidden` flag on the array items much better.

* I couldn't see what was the problem @ylecuyer ran into with `_.debounce`, I will change to `afterkeyup` and try to see if there's any problem.

**Comment by ZimGil** [comment]
Sorry for the delay had a busy few weeks,

I was able to fix the placeholder and the debounce looks fine after working with this for a few days.

I've tried to add an `isVisible` flag to the array items but I could not remove it from the list.

I've tried add either `ref.isVisialbe = true` or `ref.isVisible = ko.observable(true)`
and change it either by `isVisible = false` or `isVisible(false)` and in both cases I saw the change but it kept on showing on the list when I had `data-bind="if: isVisible` on the `li`

**Comment by ZimGil** [comment]
After working with this for a while I find the ungit process overloads the CPU from time to time

**Comment by jung-kim** [comment]
> I don't really like the auto complete solution, you'll end-up having the full list and the auto complete coveting it in some way. I like the idea of having an isHidden flag on the array items much better.

But your solution isn't using `isHidden`, you basically have copy of the array which you are against...?  I would much rather use already implemented library if possible as this would be reinventing the wheels and lacks some of the features, such as fuzzy matching or case insensitive matching.

> After working with this for a while I find the ungit process overloads the CPU from time to time

is this due to before or after this feature?

**Comment by ZimGil** [comment]
> But your solution isn't using `isHidden`, you basically have copy of the array which you are against...? I would much rather use already implemented library if possible as this would be reinventing the wheels and lacks some of the features, such as fuzzy matching or case insensitive matching.

As mentioned in my [previous comment](https://github.com/FredrikNoren/ungit/pull/1200#issuecomment-498695442), I've tried to go with the visibility flag but could not get it to work well...
I'm not sure I was doing it the correct way.

> > After working with this for a while I find the ungit process overloads the CPU from time to time
> 
> is this due to before or after this feature?

On this branch

**Comment by jung-kim** [comment]
Sorry I have been busy but every single comments adds further reasons to use auto complete library

This PR utilizes the very reason why you didn't wanted to use auto complete library because initial approach didn't work and it even overloads CPU.

**Comment by ZimGil** [comment]
I'm sorry but I won't have the time to implement this as I'm very busy for the next couple of weeks...

**Pull request closed without merging**